File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/Fibo.swift
import Foundation

func fibonacci(_ n : Int) -> Int {
	if (n < 2) {
		return n
	}
	return fibonacci(n - 1) + fibonacci(n - 2)
}

func fibonacciBU(_ n : Int) -> Int {
	if (n < 2) {
		return n
	}

	var dp : [Int] = Array(repeating: 0, count: n + 1)
	dp[0] = 0
	dp[1] = 1

	var i : Int = 2
	while (i <= n) {
		dp[i] = dp[i - 2] + dp[i - 1]
		i += 1
	}
	
	return dp[n]
}

func fibonacciBU2(_ n : Int) -> Int {
	if (n < 2) {
		return n
	}

	var first : Int = 0
	var second : Int = 1
	var temp : Int = 0

	var i : Int = 2
	while (i <= n) {
		temp = first + second
		first = second
		second = temp
		i += 1
	}
	
	return temp
}

func fibonacciTD(_ n : Int) -> Int {
	var dp : [Int] = Array(repeating: 0, count: n + 1)
	return fibonacciTD(n, &dp)
}

func fibonacciTD(_ n : Int, _ dp : inout [Int]) -> Int {
	if (n < 2) {
		dp[n] = n
		return dp[n]
	}
	
	if (dp[n] != 0) {
		return dp[n]
	}

	dp[n] = fibonacciTD(n - 1, &dp) + fibonacciTD(n - 2, &dp)
	return dp[n]
}

// Testing code.
print(fibonacci(10))
print(fibonacciBU(10))
print(fibonacciBU2(10))
print(fibonacciTD(10))

/* 
55
55
55
55
 */

File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/Fibo.ts
function fibonacci(n: number): number {
	if (n < 2) {
		return n;
	}
	return fibonacci(n - 1) + fibonacci(n - 2);
}

function fibonacciBU(n: number): number {
	if (n < 2) {
		return n;
	}

	const dp: number[] = new Array(n + 1);
	dp[0] = 0;
	dp[1] = 1;

	for (let i = 2; i <= n; i++) {
		dp[i] = dp[i - 2] + dp[i - 1];
	}

	return dp[n];
}

function fibonacciBU2(n: number): number {
	if (n < 2) {
		return n;
	}

	let first = 0;
	let second = 1;
	let temp = 0;

	for (let i = 2; i <= n; i++) {
		temp = first + second;
		first = second;
		second = temp;
	}

	return temp;
}

function fibonacciTD(n: number): number {
	const dp: number[] = new Array(n + 1);
	return fibonacciTDRecursive(n, dp);
}

function fibonacciTDRecursive(n: number, dp: number[]): number {
	if (n < 2) {
		return (dp[n] = n);
	}

	if (dp[n] !== undefined) {
		return dp[n];
	}

	dp[n] = fibonacciTDRecursive(n - 1, dp) + fibonacciTDRecursive(n - 2, dp);
	return dp[n];
}

console.log(fibonacci(10));
console.log(fibonacciBU(10));
console.log(fibonacciBU2(10));
console.log(fibonacciTD(10));
/*
55
55
55
55
*/
